presto-native-execution/presto_cpp/main/LinuxMemoryChecker.cpp

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <boost/regex.hpp>
+#include <folly/Conv.h>
+#include <folly/CppAttributes.h>
+#include <folly/File.h>
+#include <folly/FileUtil.h>
+#include <folly/Format.h>
+#include <folly/Range.h>
+#include <folly/Singleton.h>
+#include <folly/String.h>
+#include <folly/gen/Base.h>
+#include <folly/gen/File.h>
+#include <folly/gen/String.h>
+#include "presto_cpp/main/PeriodicMemoryChecker.h"
+#include "presto_cpp/main/common/Configs.h"
+
+#include <sys/stat.h>
+
+namespace facebook::presto {
+
+class LinuxMemoryChecker : public PeriodicMemoryChecker {
+ public:
+  explicit LinuxMemoryChecker(const PeriodicMemoryChecker::Config& config)
+      : PeriodicMemoryChecker(config) {}
+
+  ~LinuxMemoryChecker() override {}
+
+ protected:
+  // Current memory calculation used is inactive_anon + active_anon
+  // Our first attempt was using memInfo memTotal - memAvailable.
+  // However memInfo is not containerized so we reserve this as a
+  // last resort.
+  //
+  // Next we tried to use what docker/kubernetes uses for their
+  // calculation. cgroup usage_in_bytes - total_inactive_files.
+  // However we found out that usage_in_bytes is a fuzz value
+  // and has a chance for the sync to occur after the shrink
+  // polling interval. This would result in double shrinks.
+  //
+  // Therefore we decided on values from the memory.stat file
+  // that are real time statistics. At first we tried to use
+  // the calculation suggested by the kernel team RSS+CACHE(+SWAP)
+  // However we noticed that this value was not closely related to the
+  // value in usage_in_bytes which is used to OOMKill. We then looked
+  // at all of the values in the stat file and decided that
+  // inactive_anon + active_anon moves closest to that of
+  // usage_in_bytes
+  //
+  // NOTE: We do not know if cgroup V2 memory.current is a fuzz
+  // value. It may be better than what we currently use. For
+  // consistency we will match cgroup V1 and change if
+  // necessary.
+  int64_t systemUsedMemoryBytes() override {
+    static std::string statFile;
```

Comment:
Can we move the logic that is done one into the constructor.
Make the statFile a member of the class and initialize it in the constructor.

Only the things needed for reevaluation should be in this function.
All definitions of statics can be private members.

Reply:
Done